test/hotspot/jtreg/compiler/vectorapi/reshape/utils/VectorReshapeHelper.java

```diff
@@ -0,0 +1,356 @@
+                int expected = i < osp.vectorByteSize() ? input[i] : 0;
+                int actual = output[i];
+                Asserts.assertEquals(expected, actual);
+            }
+        }
+    }
+
+    // All this complication is due to the fact that vector load and store with respect to byte array introduce
+    // additional ReinterpretNodes, several ReinterpretNodes back to back being optimized make the number of
+    // nodes remaining in the IR becomes unpredictable.
+    @ForceInline
+    public static <T, U> void vectorRebracket(VectorSpecies<T> isp, VectorSpecies<U> osp, Object input, Object output) {
+        var outputVector = isp.fromArray(input, 0).reinterpretShape(osp, 0);
+        var otype = osp.elementType();
+        if (otype == byte.class) {
+            ((ByteVector)outputVector).intoArray((byte[])output, 0);
+        } else if (otype == short.class) {
+            ((ShortVector)outputVector).intoArray((short[])output, 0);
+        } else if (otype == int.class) {
+            ((IntVector)outputVector).intoArray((int[])output, 0);
+        } else if (otype == long.class) {
+            ((LongVector)outputVector).intoArray((long[])output, 0);
+        } else if (otype == float.class) {
+            ((FloatVector)outputVector).intoArray((float[])output, 0);
+        } else if (otype == double.class) {
+            ((DoubleVector)outputVector).intoArray((double[])output, 0);
+        } else {
+            throw new AssertionError();
+        }
+    }
+
+    public static <T, U> void runRebracketHelper(VectorSpecies<T> isp, VectorSpecies<U> osp) throws Throwable {
+        var random = RandomGenerator.getDefault();
+        String testMethodName = VectorSpeciesPair.makePair(isp, osp).format();
+        var caller = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE).getCallerClass();
+        var testMethod = MethodHandles.lookup().findStatic(caller,
+                    testMethodName,
+                    MethodType.methodType(void.class, isp.elementType().arrayType(), osp.elementType().arrayType()))
+                .asType(MethodType.methodType(void.class, Object.class, Object.class));
+        Object input = Array.newInstance(isp.elementType(), isp.length());
+        Object output = Array.newInstance(osp.elementType(), osp.length());
+        long ibase = UnsafeUtils.arrayBase(isp.elementType());
+        long obase = UnsafeUtils.arrayBase(osp.elementType());
+        for (int iter = 0; iter < INVOCATIONS; iter++) {
+            for (int i = 0; i < isp.vectorByteSize(); i++) {
+                UnsafeUtils.putByte(input, ibase, i, random.nextInt());
+            }
+
+            testMethod.invokeExact(input, output);
+
+            for (int i = 0; i < osp.vectorByteSize(); i++) {
+                int expected = i < isp.vectorByteSize() ? UnsafeUtils.getByte(input, ibase, i) : 0;
+                int actual = UnsafeUtils.getByte(output, obase, i);
```

Comment:
When `MemorySegment` previews we can remove the use of unsafe accesses, since we can view the array as a segment. Same i think also applies to direct VH accessors.